scraper/src/libretexts2zim/client.py

```diff
@@ -0,0 +1,135 @@
+import datetime
+import re
+from collections.abc import Callable
+
+import requests
+from bs4 import BeautifulSoup, NavigableString
+from pydantic import BaseModel
+
+from libretexts2zim.constants import logger
+
+HTTP_TIMEOUT_SECONDS = 15
+
+
+class LibreTextsParsingError(Exception):
+    pass
+
+
+class LibreTextsShelve(BaseModel):
+    title: str
+    content_url: str
+    image_url: str
+
+
+class LibreTextsHome(BaseModel):
+    welcome_text_paragraphs: list[str]
+    shelves: list[LibreTextsShelve]
+    welcome_image_url: str
+
+
+class LibreTextsMetadata(BaseModel):
+    """Metadata about a course."""
+
+    # Human readable name for the course.
+    name: str
+
+    # URL prefix for the course, e.g. for Geosciences which is at
+    # https://geo.libretexts.org/, the slug is `geo`
+    slug: str
+
+    def placeholders(
+        self, clock: Callable[[], datetime.date] = datetime.date.today
+    ) -> dict[str, str]:
+        """Gets placeholders for filenames.
+
+        Arguments:
+          clock: Override the default clock to use for producing the "period".
+        """
+
+        return {
+            "name": self.name,
+            "slug": self.slug,
+            "clean_slug": re.sub(r"[^.a-zA-Z0-9]", "-", self.slug),
+            "period": clock().strftime("%Y-%m"),
+        }
+
+
+class LibreTextsClient:
+    """Utility functions to read data from libretexts."""
+
+    def __init__(self, library_slug: str) -> None:
+        """Initializes LibreTextsClient.
+
+        Paremters:
+            library_url: Scheme, hostname, and port for the Libretext library
+                e.g. `https://geo.libretexts.org/`.
+        """
+        self.library_slug = library_slug
+
+    @property
+    def library_url(self) -> str:
+        return f"https://{self.library_slug}.libretexts.org/"
+
+    def _get_text(self, url: str) -> str:
+        """Perform a GET request and return the response as decoded text."""
+
+        logger.debug(f"Fetching {url}")
+
+        resp = requests.get(
+            url=url,
+            allow_redirects=True,
+            timeout=HTTP_TIMEOUT_SECONDS,
+        )
+        resp.raise_for_status()
+
+        return resp.text
+
+    def get_home(self) -> LibreTextsHome:
+        home_content = self._get_text(self.library_url)
+
+        soup = _get_soup(home_content)
+        return LibreTextsHome(
+            welcome_text_paragraphs=_get_welcome_text_from_home(soup),
+            welcome_image_url=_get_welcome_image_url_from_home(soup),
+            shelves=[],
+        )
+
+
+def _get_soup(content: str) -> BeautifulSoup:
+    return BeautifulSoup(content, "html.parser")
+
+
+def _get_welcome_image_url_from_home(soup: BeautifulSoup) -> str:
+    branding_div = soup.find("div", class_="LTBranding")
+    if not branding_div:
+        raise LibreTextsParsingError("<div> with class 'LTBranding' not found")
```

Comment:
The whole client is tested with integration tests, so should the DOM change we will be warn sooner than later (especially since I now think integration tests should be run daily as well, just opened https://github.com/openzim/libretexts/issues/25).

These various checks you saw are important to please the type checker (not really a good reason) and to have a precise error message.

I also think that checking DOM integrity at the moment we use it is important to ensure we cover all cases and checks consistency. Having to ensure consistency between checking code which is somewhere and "using" code which is somewhere else is never easy.

This is also OK because this code is ran fairly early in the scrape logic, so it will make the scraper fail early as well. In Wikihow I suspect we want to check the DOM early also to avoid running the scraper too long before realizing that the DOM is finally wrong and we cannot proceed. Isn't it?

I think we do not need early checks of the DOM for now, but it is true that it might change in the future. More something to keep in mind from my PoV, but anyway thank you for the good advice.

Reply:
OK, you convinced me but I'm afraid the early testing might not stand in with time. This is an early PR with limited output.
The daily tests in CI is a good move that we should probably include in our convention.